compiler/vm/vmdef.nim

```diff
@@ -90,27 +90,138 @@ type
+    slotTempPerm      ## slot is temporary but permanent (hack)
+
+  AtomKind* = enum
+    akInt
+    akFloat
+    akString # TODO: maybe even collapse akString into akOrdinalArray?
+    akOrdinalArray # Space efficient storing of float/int arrays
+    akArray # Used for both seqs and fixed size arrays by vmgen
+    akSet
+    akRef
+    akPtr
+    akPointer ## A raw pointer points to inside akString's string or akOrdinalArray/akArray's rawArrayVal
+    akObject
+    akPNode
+    akCallable
+
+  CallableKind* = enum
+    ckDefault ## A referenc to a VM function
+    ckClosure
+    ckCallback ## A VmCallback
+    ckFFI
+    ckTemplate
+
+  # XXX: design is temporary
+  VmFunctionObject* = object ## An object storing a callable
+    case kind*: CallableKind
+    of ckDefault:
+      prcSym*: PNode # TODO: needs to be a `PSym`
+    of ckClosure:
+      closureEnv*: GlobalRef ## A reference to the closure environment
+      closureTarget*: int
+    of ckCallback:
+      cbOffset*: int
+    of ckFFI:
+      ffiPrcSym*: PSym
+    of ckTemplate:
+      templateSym*: PSym
+
+
+  OrdinalArrayKind* = enum
+    oakUnsigned
+    oakSigned
+    oakFloat32
+    oakFloat
+
+  VmObjectAtom* = object
+    case kind*: AtomKind
+    of akInt:
+      # XXX: is it okay to use this for unsigned integer too?
+      intVal*: BiggestInt
+    of akFloat:
+      floatVal*: BiggestFloat
+    of akString:
+      strVal*: string
+    of akOrdinalArray:
+      rawArrayVal*: seq[byte]
+      elemSize*: int8 # 1, 2, 4 or 8
+      oaKind*: OrdinalArrayKind # TODO: encode `oaKind` and `elemTyp` into a single int8
+    of akArray:
+      arrayElements*: seq[VmObjectAtom]
+      elemTyp*: PType
+    of akSet:
+      # Theoretically sets could use nim `set`s for representing data. But that would more
+      # or less require the use of an heap-allocated structure to contain the set and also casing
+      # between set[uint8] and set[uint16] since otherwise `VmObjectAtom` would explode to the size
+      # of 8192 byte (sizeof(set[uint16]))!
+
+      # XXX maybe use `ptr IntSet`? Can't use IntSet with automatic storage duration since
+      # sizeof(IntSet) is also huge (smaller than set[uint16] however)
+
+      # XXX should `setElems` be ordered? Would speed up contains checks but require sorting on modification
+      # XXX `setElem` could also be run-length encoded for speed and memory efficiency...
+      setElems* : seq[uint16]
+    of akPtr:
+      ptrTyp*: PType 
+      # XXX should we keep a distinction between pointer to global and poiunter to inside register?
+      ptrVal* : ptr VmObjectAtom
+    of akRef:
+      # TODO: should we use a ptr?
+      globalIndex*: GlobalRef ## An index into
+    of akPointer:
+      pointerTyp*: PType
+      pointerVal*: pointer
+    of akObject:
+      # TODO: what about type hooks (=copy, =sink, =destroy)? Original vm implementation seems to 
+      # have completely ignored these
```

Comment:
I think we should have vmgen emit the locations where the hooks would be run.

Reply:
What do you mean by location?

Reply:
`=copy` and `=sink` can be automatically emitted by the compiler (it already does that for native code backends), so I don't think we have to worry too much.

What I'm not sure of is whether the phase that gets these stuff ticking (`injectdestructors`) is used to transform the code for vmgen.

Reply:
As far as I can tell, the code is simply transformed with `transf.transformBody` by `vmgen` prior to code gen right now. `injectdestructors` is not used.

Problems would arise with `ref` and `seq` types.

`ref` handling is fully done by the execution engine right now, so `vmgen` has no clue to where and when their lifetime ends. `liftdestructors` seems to be responsible for the injection of ARC/ORC reference handling and maybe we can leverage this for `vmgen` too. There are a whole lot of codegen procs that'd need to be handled then though (`nimIncRef`, `nimDecRefLast`, `nimIncRefCyclic`, etc). I'm not sure if this is the right approach.

All `seq` related functionality is handled via dedicated instructions, meaning that the workings of seq handling are completely opaque to `vmgen`. A few ideas:
- Make `seq`s a non-magic type (very likely a complex endeavor and one I'd not attempt as part of this PR, but could be interesting in the future)
- Leave the seq opcodes intact but remove dedicated atom support for them from the execution engine. Use `seqs_v2` for both the storage implementation and logic. Opaquely transform seq instructions to a VM procedure call - that is, run `vmgen` over `setLen`, `shrink`, etc. and execute them directly as guest code
- Same as the above, but remove seq opcodes and let `vmgen` emit calls to `seqs_v2` procs directly.
- Nim type information needs to be carried around anyway, so VM `seqs` know their elements' type. Whenever elements are to be copied/moved/destroy, check for the existence of a corresponding lifetime-hook and run it as guest code for the required elements. Something similar to this could also be employed for `ref` type destructors

Removing dedicated `seq` support from the VM would make it less complex, but usage of `seq`s would also be slower due to each operation running as guest code.

The `=trace` hook can probably only be realistically supported if we make use of `injectdestructors`/`liftdestructors` somehow.